bluepysnap/query.py

```diff
@@ -0,0 +1,159 @@
+    """
+    if ids is None:
+        return np.full(len(data), fill_value=True)
+    mask = np.full(len(data), fill_value=False)
+    valid_ids = pd.Index(utils.ensure_list(ids)).intersection(data.index)
+    mask[valid_ids] = True
+    return mask
+
+
+def _circuit_mask(data, population_name, queries):
+    """Handle the population, node ID queries."""
+    populations = queries.pop(POPULATION_KEY, None)
+    if populations is not None and population_name not in set(utils.ensure_list(populations)):
+        ids = []
+    else:
+        ids = queries.pop(NODE_ID_KEY, queries.pop(EDGE_ID_KEY, None))
+    return queries, _positional_mask(data, ids)
+
+
+def _properties_mask(data, population_name, queries):
+    """Return mask of IDs matching `props` dict."""
+    unknown_props = set(queries) - set(data.columns) - ALL_KEYS
+    if unknown_props:
+        return np.full(len(data), fill_value=False)
+
+    queries, mask = _circuit_mask(data, population_name, queries)
+    if not mask.any():
+        # Avoid fail and/or processing time if wrong population or no nodes
+        return mask
+
+    for prop, values in queries.items():
+        prop = data[prop]
+        if np.issubdtype(prop.dtype.type, np.floating):
+            v1, v2 = values
+            prop_mask = np.logical_and(prop >= v1, prop <= v2)
+        elif isinstance(values, str) and values.startswith('regex:'):
+            prop_mask = _complex_query(prop, {'$regex': values[6:]})
+        elif isinstance(values, Mapping):
+            prop_mask = _complex_query(prop, values)
+        else:
+            prop_mask = np.in1d(prop, values)
+        mask = np.logical_and(mask, prop_mask)
+    return mask
+
+
+def traverse_queries_bottom_up(queries, node_function):
+    """Traverse queries tree from leaves to root, left to right.
+
+    Args:
+        queries (dict): queries
+        node_function (function): function to execute on each node of queries tree in traverse order
+    """
+    for key in list(queries.keys()):
+        if key in {OR_KEY, AND_KEY}:
+            for subquery in queries[key]:
+                traverse_queries_bottom_up(subquery, node_function)
+        elif isinstance(queries[key], Mapping):
+            if VALUE_KEYS & set(queries[key]):
+                if not set(queries[key]).issubset(VALUE_KEYS):
+                    raise BluepySnapError("Value operators can't be used with plain values")
+            else:
+                traverse_queries_bottom_up(queries[key], node_function)
+        node_function(queries, key)
+
+
+def get_properties(queries):
+    """Extracts properties names from `queries`.
+
+    Args:
+        queries (dict): queries
+
+    Returns:
+        set: set of properties names
+    """
+
+    def _collect(_, query_key):
+        if query_key not in ALL_KEYS:
+            props.add(query_key)
+
+    props = set()
+    traverse_queries_bottom_up(queries, _collect)
+    return props
+
+
+def operator_mask(data, population_name, queries):
```

Comment:
I think the name is not correct anymore (and maybe even before when it was in the NodePopulation class). Something like : 
resolve_ids or apply_query
could be better no ?

It would be more elegant also to have : 
def resolve_ids(population, query):
     data = population.get()
     population_name = population.name
or we can decide how to access a "public population.data". 

One problem I have right now is : how can we handle the data argument for the edges ? We will not be able to have a dataframe in memory. Do you want to use a dask dataframe ?

Reply:
> I think the name is not correct anymore (and maybe even before when it was in the NodePopulation class).

`resolve_ids` sounds good. I kept old names so the review is easier.

> One problem I have right now is : how can we handle the data argument for the edges ? We will not be able to have a dataframe in memory. Do you want to use a dask dataframe ?

I planned to use a DataFrame that contains only properties in the query. It is easy to do with `query.get_properties` function.